src/Runtime/Core/Maths/BranchAndBound.cs

```diff
@@ -92,61 +94,112 @@ public static Vector Search(Region bounds, Func<Vector, double> Evaluate, Func<R
+                            }
+                        }
                     }
-                    else
-                    {
-                        foundSplit = true;
+                    if (splitDim < 0)
                         break;
-                    }
                 }
-                if (!foundSplit)
+                else
                 {
-                    break;
+                    // Find a dimension to split on.
+                    splitDim = Rand.Int(dim);
+                    bool foundSplit = false;
+                    for (int i = 0; i < dim; i++)
+                    {
+                        if (!DimensionCanSplit(splitDim))
+                        {
+                            splitDim++;
+                            if (splitDim == dim)
+                                splitDim = 0;
+                        }
+                        else
+                        {
+                            foundSplit = true;
+                            break;
+                        }
+                    }
+                    if (!foundSplit)
+                    {
+                        break;
+                    }
                 }
 
                 // split the node
                 double splitValue = midpoint[splitDim];
+                if(Debug)
+                    Trace.WriteLine($"splitting dimension {splitDim}");
                 if (region.Upper[splitDim] != splitValue)
                 {
-                    Region leftRegion = new Region(region);
-                    leftRegion.Upper[splitDim] = splitValue;
-                    addRegion(leftRegion, splitDim);
+                    if(leftRegion == null)
+                    {
+                        leftRegion = new Region(region);
+                        leftRegion.Upper[splitDim] = splitValue;
+                        upperBoundCount++;
+                        upperBoundLeft = GetUpperBound(leftRegion);
+                    }
+                    addRegion(leftRegion, splitDim, upperBoundLeft);
                 }
                 if (region.Lower[splitDim] != splitValue)
                 {
-                    Region rightRegion = new Region(region);
-                    rightRegion.Lower[splitDim] = splitValue;
-                    addRegion(rightRegion, splitDim);
+                    if(rightRegion == null)
+                    {
+                        rightRegion = new Region(region);
+                        rightRegion.Lower[splitDim] = splitValue;
+                        upperBoundCount++;
+                        upperBoundRight = GetUpperBound(rightRegion);
+                    }
+                    addRegion(rightRegion, splitDim, upperBoundRight);
                 }
             }
-            //Trace.WriteLine($"BranchAndBound.Search upperBoundCount = {upperBoundCount}");
+            if(Debug)
+                Trace.WriteLine($"BranchAndBound.Search upperBoundCount = {upperBoundCount}");
```

Comment:
Consider Debug.Writeline

Reply:
Debug.Listeners does not exist in .NET Standard, which makes Debug.WriteLine effectively useless.